### ISSUE TITLE
PROPOSAL: Kopia Telemetry Collection

### ISSUE THREAD

**Issue by jkowalski** [kopia/kopia#2486]
## Summary

In order to better understand how people use Kopia we’d like to add telemetry feature which will collect and aggregate some basic information about installations and usage in a privacy-preserving manner. The data collected will be used to prioritize feature development and bug fixes.

This document proposes the scope of data collection to get community consensus on what kind of information is ok to collect. Please post feedback as comments and reactions on this issue.

## Goals

We will collect minimal amount of non-personally-identifying information to be able to answer the following questions:

* Which versions of Kopia are active in the wild
* What’s the respective usage of each version
* How many repositories are active?
* How big are the repositories, how many users typically share the repository, etc.
* What is the age and size of the repositories
* What providers are being used?
* Which operating systems and CPU architectures are most commonly used
* What’s the ratio of UI vs CLI usage

All collected information will be made public in the form of public dashboards on https://kopia.io/ showing trends over time and charts showing respective usage percentages.

# Collection Method

Starting with v0.13.0 we will instrument Kopia clients (both CLI and UI) to `POST` the following information to central endpoint https://telemetry.kopia.io/ping no more than once every 24 hours from each client. The mechanism will be very similar to one currently used for upgrade checks.

```
{
  “repositoryInfo”: {
    “createDate”: “2022-10-05”,
    “repositoryUniqueId”: “sha256(<unique-repository-id-assigned-at-creation-time>)”
  },
  “clientInfo”: {
    “randomConnectionId”: “<random-64-bit-id-assigned-per-connection-to-repository>”,
    “connectionDate”: “2022-10-05”,
    “version”: “0.13.0”,
    “os”: “windows|darwin|linux|freebsd|openbsd”,
    “arch”: “amd64|arm64|arm”,
    “ui”: true|false,
    “isMaintenanceOwner”: true|false,
  },
  “formatInfo”: {
    “provider”: “s3|b2|gcs|filesystem|...”,
    “hash”: “<algorithm>”,
    “encryption”: “<algorithm>”,
    “splitter”: “<algorithm>”,
    “formatVersion”: <number>,
    “indexFormatVersion”: <version>,
  },
  “stats”: {
    “repositorySizeGB”: <number>,
    “numSnapshots”: <number>, // number of snapshots in the repository
    “numSources”: <number>,   // number of unique username@hostname:/path
    “numPolicies”: <number>,  // number of policies
    “numUsers”: <number>,     // number of users sharing the repository
    “numHosts”: <number>,     // number of hosts sharing the repository
  },
}
```

## Opt-Out Method

Collection will be enabled by default and clearly explained to the users giving them instructions for opting out at the moment of connection. The first ping will be sent no sooner than 24 hours after establishing connection to a repository.

To opt out, the clients can:

* Set environment variable KOPIA_TELEMETRY=false
* Pass `--no-telemetry` flag on kopia invocation
* Uncheck a checkbox in the kopia UI.

## Data Storage

All collected data will be aggregated (for example on a daily, weekly and monthly basis) and individual ping records will be discarded after no more than 90 days with only aggregated counters stored beyond that.

**Comment by icepic** [comment]
I would prefer an opt-in. I will probably enable it on various computers that does use kopia, but I think there will be a backlash if someone misses it, no matter how visible it is. I get the idea to make it very visible and so on, but historically, opt-out seems to cause more issues with peoples perception than it is worth. At least start out with opt-in, and make it visible that you would prefer for people to allow this.

**Comment by jkowalski** [comment]
I hear you, I also prefer opt-in whenever given that choice, but the problem in this particular case is we'll never get any data about CLI usage (having to pass a flag which has no visible effect would lead to nobody doing it, especially folks who use scripts or containerized kopia).

For the UI we could and should present the choice prominently in the UI before allowing people to use it for the first time, but I would still like to preselect the telemetry option and allow folks to uncheck it, the same way that most software does it today.

Right now we're flying completely blind, for example when it comes to answering the question like how many people are on repostitory format v1 vs v2 and is it worth investing in migration support or not.

**Comment by bbccdd** [comment]
Great initiative! 

Indeed, knowing how kopia is used may very well guide development, serve existing, and attract new users. I strongly agree with privacy respecting (anonymous: meaning no individual persons can be traced) way _by default_.

To serve anyone who prefers opt-in, you may consider clearly supplying a notice statement in the next update that anonymous usage statistics will be collected to improve the development of kopia in a (near) future release and how to disable that. If you are planning to release v0.13 next, you may consider adding a v0.12.x release in between that supports the --no-telemetry flag amongst other updates.

NB Perhaps the 'anonymous usage statistics' can be reported back publicly in a simple web interface. For inspiration: https://usage-reporter.duplicati.com. These insights are also useful for users of kopia, e.g., when determining the splitter, hash, crypto functions based on popularity and the assumed larger interest in long-term support of those choices.

**Comment by budachst** [comment]
As mentioned right at the top, the information transferred will not contain any PII, so am fine with the opt-out. I guess, our use of Kopia is more on the excessive side, so I want our data to be considered, when it comes to what features and bug fixes will be prioritized. ;)

**Comment by xxxliqu1dxxx** [comment]
Trying not to overthink this, but if I may ask, what is the use case of collecting repositoryUniqueId?

Isn't there technically a way to tie repositoryUniqueId back to a user by cross checking the repositoryUniqueId with a github post, screenshot and/or attachment which would contain the output of "repo status" for instance? If so that becomes a piece of information that could personally identify a user.

I understand the general use case of collecting some information to improve the tool, but not sure I understand the benefit of collecting individual repo ids. I suppose the same could be said for randomConnectionId...

**Comment by jkowalski** [comment]
Good point. We should be collecting hashes of unique ids not the ids themselves. The only purpose is to get unique repository count. 

As for the other field, its purpose is to understand how many unique kopia connections to repositories there are and detect cases where clients get upgraded with and without reconnecting. 

Random id could be replaced with connection time stamp which would give us duration as well.

**Comment by basldfalksjdf** [comment]
As long as PII is not being collected, this seems fine to me. I also don't see as opt-in being a viable option. If the goal of this is to gather data that guides Kopia development, it needs to be opt-out, because otherwise not much data will be gathered.

**Comment by ntolia** [comment]
As a kopia user and long-term supporter that wants to see the project grow further, I think this is a great idea. We will not be able to figure out what is being heavily used and what use cases to support without this. As others have stated, this has to be opt-out for us to gather meaningful data.

Can we also gather data on the amount of metadata kopia is managing (number of files, blocks, packs, indices, etc.)? That and the number of snapshots (already in there), will give us a good idea of the scale for each snapshot and the metadata overhead we have. I wonder how many folks are pushing kopia in directions we don't expect.

**Comment by jkowalski** [comment]
How about the following plan:

1. We will develop this feature in 0.13.0 timeframe, but it will be disabled by default (opt-in mode), so that only developers and users interested in helping out will be able manually trigger it while we develop data collection endpoint, storage and and dashboards.

2. When ready, we will switch the default for the `unstable` releases to enabled (opt-out mode), while `stable` releases will remain as opt-in. This will start collecting some data from a fleet of unstable deployments to prove aggregation and dashboarding. Currently only about 0.1% of users are on `unstable` releases.

3. We will run in this mode for several weeks to gather some data to populate dashboards.

4. We will release 0.13.0 in opt-in mode for both CLI and UI but with a notice that in 0.14.0 we will make telemetry collection on by default, giving folks time to adjust scripts if they don't want to share the stats.

**Comment by ntolia** [comment]
@jkowalski Are we adding too much overhead and time to the process? Assuming with we go with something like Jitsu, the SDK integration for shipping telemetry should be quite simple and stable. I would just develop with it being on (easier to test the happy path) and we can always throw away data if there are any initial bugs.

Would love to see what others think.

**Comment by icepic** [comment]
I was thinking there might possibly exist an opt-out (or opt-in for that matter) that exists as an object in the repository, somewhat like the "if you dont want the auto-updater, touch $HOME/.config/kopia/update-check.json" or what that message is.
This way, a client that gets attached later on to my "super-secret-dont-want-to-leak-any-info" repo will not send info about it just because someone forgot --dont-send-usage-stats=true at some point in the future.

**Comment by stpr-dev** [comment]
I agree that it should be opt-in with user consent. After all, this is a secure backup utility, so there is some expectation of privacy. I would also like to add that it might be better to use broad ranges instead of concrete numbers - instead of collecting "User X has 2 repos (123 GB and 678GB), etc...", do ranges (User X as < 10 repos, range between 100GB-1TB). This is just to ensure that users are more comfortable sharing knowing that it ranges and nothing concrete.

**Comment by jkowalski** [comment]
@stpr-dev 

I love the idead of bucketizing values into broad ranges where it makes sense and makes total sense for counters.

As for your other comment, there is really no notion of "User X" (and we're not interested in that). Given he proposed scheme it's impossible to tell if a user has two repositories, or 100 since we don't plan to identify users at all - only repositories using stable IDs and individual connections using random identifiers that change on every connection.

What we'll be able to say is things like `repository <random-id> has 20 active unique connections in the last 7 days". 

When we aggregate this information, we will only keep daily facts like:

```
On 2022-10-07:

* X repositories had 1 connection in the last 7 days
* X repositories had 2-5 connections in the last 7 days
* X repositories had 6-10 connections in the last 7 days
* X repositories had 10-50 connections in the last 7 days
* X repositories had 51-100 connections in the last 7 days
* X repositories had 100-500 connections in the last 7 days
* X repositories > 500 connections in the last 7 days

* Y repositories were between 100MB and 1GB
* Y repositories were between 1GB and 10GB
* Y repositories were between 10GB and 50GB
* Y repositories were between 50GB and 100GB
* Y repositories were between 100GB and 200GB
* Y repositories were between 200GB and 300GB
* Y repositories were between 300GB and 400GB

```

There's really not much of ability to track users with the data we collect. If you think there is something objectionable, let's talk about removing it - now is the time to speak up.

Also I had some additional conversations offline and here are some additional things we're going to do:

- whenever a TTY / interactive user session (as opposed to batch) is detected, we'll ask for permissions to send telemetry on `repository create` and `repository connect`

- we will develop repository-level policy where repository administrator can specify per-connection policy with options like:

  - always allow telemetry
  - always disallow telemetry
  - ask for individual consent, default to allow
  - ask for individual consent, default to disallow

- we will develop a way to override the telemetry endpoint (via policies) so that administrators can inspect the traffic and forward it upstream

- in the interest of complete transparency we're planning to open source all parts of the telemetry solution including source code for the client that sends the pings, collector that receives them, aggregator and the dashboard. Aggregated data will be made public in the raw form too.

**Comment by stpr-dev** [comment]
Ah, I see - the example makes it clear what your intent is. I think the key points that I can summarise from this thread would be:
- Planned clear and explicit UI/CLI message asking permission for telemetry during repo creation/migration.
- Clearly indicating the purpose (help with feature developments).
- Indicate what is and isn't collected (not collecting personal file information, just an overview of the setup).
- Letting users know that they can use policy/command/UI button to opt-in or opt-out at any time.
- Data will be made available for users to see (raw and fancy).

A couple of points to consider:
- It might be a good plan to document this in plain English explaining all of these that a user can visit at any time.
- Is https://telemetry.kopia.io/ping going to be the planned ping server? If so, I think that it might be good to let users know where the website is hosted (some people appear to strongly object to some providers).

Other than that, given the information, I think I don't have any major objections.

**Comment by xxxliqu1dxxx** [comment]
> * Planned clear and explicit UI/CLI message asking permission for telemetry during repo creation/migration.

Not only during repo creation/migration, but also at the installalation/update stage. There needs to be an tangible checkbox to "opt-in" or to "opt-out" at the update level. An end user could opt to stay on a certain version to avoid all telemetry, but the first update that enables that needs to have this as a checkbox for sure.

> * Indicate what is and isn't collected (not collecting personal file information, just an overview of the setup).

I would actually display a sample of what would be sent based on the data on the repo. Using actual data to sample, this way the user would understand based on "their data" vs "generic sample".

**Comment by stpr-dev** [comment]
> Not only during repo creation/migration, but also at the installalation/update stage. There needs to be an tangible checkbox to "opt-in" or to "opt-out" at the update level. An end user could opt to stay on a certain version to avoid all telemetry, but the first update that enables that needs to have this as a checkbox for sure.

I am not a 100% sure how this will work with, for instance, linux package managers. I can see, for instance, using the auto-updater with the UI in Windows that it might be able to display prompt, but what if I do `yay -S kopia` on Arch? I am not sure that it will work out as smoothly - It might be tricky to consider these cases. Just as a thought, perhaps when launching *after an update*, a message can be prominently displayed asking for user consent. Should they refuse to accept it, then they can rollback an update and stick with the old one if they really wanted to.

> I would actually display a sample of what would be sent based on the data on the repo. Using actual data to sample, this way the user would understand based on "their data" vs "generic sample".

Good idea! Show them exactly what is transmitted (raw and fancy) so that they know *exactly* what is transmitted. Although the source code is planned to be public, it is good to show end users who are not familiar or comfortable with navigating the source themselves.

**Comment by xxxliqu1dxxx** [comment]
> Good idea! Show them exactly what is transmitted (raw and fancy) so that they know _exactly_ what is transmitted. Although the source code is planned to be public, it is good to show end users who are not familiar or comfortable with navigating the source themselves.

Yes. exactly!

**Comment by githubbapoopa** [comment]
What about getting the "simple" stuff right in the first place?

Doc, doc, docs. Wikis. Portable versions...

A professional backup software should never have any form of telemetry. That's a wet dream.

If you do this kopis is officialy dead for pro usage imho.

Get your prioriities right.

Cheers

**Comment by stpr-dev** [comment]
> Doc, doc, docs. Wikis. Portable versions...

Docs are being constantly updated, and they will get better over time. Portable versions already exist? I am not sure what you mean, the cli is a single binary that operates on its own; and there are zip files for the UI version (also AppImages).

> A professional backup software should never have any form of telemetry in the first place.

I think that a backup software shouldn't have *mandatory* telemetry. However, the proposal is clearly stating plans for:

- User consent and policy/settings to fully exclude any telemetry
- Not collecting any identifiable information (and is also giving access to the information that is being sent to the server)
- Fully Open Source, including the server code so that you can inspect what is being sent

Given that Kopia is a free product mainly developed by a few individuals, it is not unreasonable for the developers to ask what where they should prioritize their efforts. There is no other good alternative: most users won't file GitHub feature requests or issues, and/or won't necessarily participate in the forums. If you still have issues with that, perhaps a compile time flag can be included to completely exclude telemetry module? That way, anyone can compile your own binary without any of the telemetry.

**Comment by jkowalski** [comment]
I've been thinking more about it and here's another refinement of the idea to make this completely opt-in:

- we will have per-repository setting whether user has given consent to sharing telemetry: `enabled|disabled|unknown`
- for new users we will explicitly ask for user consent when creating a new repository (see below):
  - if repository creation is interactive (session has TTY), we will ask for consent in the terminal, right after asking for password
  - in the UI we will present a checkbox before first use
  - if creation is in a batch script or when no TTY is available, we will __require__ passing `--share-anonymous-statistics=enabled|disabled` which can alternatively be provided using `KOPIA_SHARE_ANONYMOUS_STATISTICS` environment variable
- for existing users if the consent is unknown we will try to ask user whenever the user is using Kopia interactively (as opposed to batch script). Until consent is obtained we will not share anything:

For example:

```
$ kopia snapshot list ...
<list output>
<consent-prompt>
```

The consent prompt would require making an explicit choice (there's no default) and will be something like:

```
Kopia would like to request your permission to share anonymous statistics about this repository:

[1] I'm ok to share anonymous data. This information only includes basic statistics, 
such as repository size, version and format information, etc.  and never any
personally-identifiable information.

[2] I prefer not to share.

See https://kopia.io/docs/telemetry and https://kopia.io/privacy for more information.
The information you provide will be used to improve Kopia and we really appreciate your help.

Enter 1 or 2: 
```

**Comment by xxxliqu1dxxx** [comment]
Thanks for adding details Jarek. If someone decides to "I prefer to not share", will they be prompted every time to make that decision? Alternatively if someone wants to change their mind and want to start sharing, then they would have to set the flag with a parameter?

**Comment by stpr-dev** [comment]
> The consent prompt would require making an explicit choice (there's no default) and will be something like:
> 
> ```
> Kopia would like to request your permission to share anonymous statistics about this repository:
> 
> [1] I'm ok to share anonymous data. This information only includes basic statistics, 
> such as repository size, version and format information, etc.  and never any
> personally-identifiable information.
> 
> [2] I prefer not to share.
> 
> See https://kopia.io/docs/telemetry and https://kopia.io/privacy for more information.
> The information you provide will be used to improve Kopia and we really appreciate your help.
> 
> Enter 1 or 2: 
> ```

As @xxxliqu1dxxx, this might be a good time to show a sample of *exactly* what is being sent, like:

```
Kopia would like to request your permission to share anonymous statistics about this repository. The following anonymous basic information will be shared:

** INSERT FANCY OUTPUT HERE**

[1] I'm ok to share this anonymous data. This information only includes basic statistics, 
such as repository size, version and format information, etc.  and never any
personally-identifiable information.

[2] I prefer not to share.

See https://kopia.io/docs/telemetry and https://kopia.io/privacy for more information.
The information you provide will be used to improve Kopia and we really appreciate your help.

**Please note that you can opt-in or opt-out at any time.**

Enter 1 or 2: 
```

As you had mentioned, the flag could enable or revoke consent at anytime.

**Comment by jkowalski** [comment]
@xxxliqu1dxxx no, it's going to ask only if we don't know the answer, if the answer is known, we will never ask.

I'm thinking we will also need a way to change the decision:

```
$ kopia repository telemetry status
$ kopia repository telemetry enabled
$ kopia repository telemetry disable
```

**Comment by jkowalski** [comment]
Btw, "ask me in a week" may be an interesting possibility - for users who are new to Kopia and want to try using it first.

**Comment by basldfalksjdf** [comment]
While I understand some people's reservations regarding opt-out, I still think it needs to be opt-out rather than opt-in if there is going to be any kind of meaningful (yet anonymous) data collection that will help guide Kopia's development.

If we go the opt-in route, not only will there be very few users that opt-in (just by nature, not necessarily because something nefarious is going on) but also the data will be biased and skewed towards more involved and/or more tech savvy Kopia users. It will not provide very meaningful insights into Kopia usage and may even mislead Kopia development, as it will only represent usage by a select type of Kopia user.

As long as the data collection is transparent and anonymous, I suspect there is a silent majority that is okay with opt-out, even if the loud minority still want opt-in. After all, we all want Kopia to be around for the long-term, and it just cannot continue to be developed blindly (figuratively speaking). I also suspect that the same people who are pushing for opt-in would much rather not have these features in Kopia to begin with, so not sure why this feature should be unnecessarily catered to their demands.

The point of adding anonymous usage data gathering is to get a better understanding of how people use Kopia so that it can guide development. We may as well not add this feature if we are going to handicap it by design.

(It goes without saying, by "we" I mean the devs who will actually put the work into adding this.)

**Comment by basldfalksjdf** [comment]
> What about getting the "simple" stuff right in the first place?
> 
> Doc, doc, docs. Wikis. Portable versions...

Not everyone has the ability to code, so we all cannot contribute to Kopia that way. I get that. But surely anyone can contribute to docs/wikis? If you think that the docs/wikis are not "right", how about fixing em instead of complaining?

> A professional backup software should never have any form of telemetry in the first place.
> 
> That's a wet dream.
> 
> That's not professional it's the opposite.

Hard disagree. Any "professional" software needs telemetry -- to help identify issues and guide development. That is just the nature of software development.

I agree that the telemetry should be consistent with the mission of the program. In Kopia's case, privacy is a huge emphasis, so the telemetry should be accordingly -- no private information collected and everything should be appropriately anonymized. But it really is a must. 

> If you do this kopia is officially dead for pro usage imho.

I suppose this is always a risk, but I highly doubt it. Especially as pros could easily, you know, opt out...

> Get your priorities right.
> 
> Cheers

...

**Comment by typpos** [comment]
@jkowalski 
Can you give some concrete examples how the statistics obtained from a data collection exercise will result in actual development actions? The "goals" listed in your original post are data collection goals, not project goals.

A key concern about telemetry-driven development is that it re-enforces workarounds and current behaviour, not what users want/need to do. For example, I may need to use the CLI version because it can create all snapshots in a single call (snapshot create --all), whereas kopia UI does not. Telemetry would tell you that I prefer CLI over UI and incorrectly prioritize development that way. Similarly, if my repositories are smaller than a given size, then this may be an existing performance constraint I am working around, not a desired size.

**Comment by cristihcd** [comment]
I think opt-out should be the default, otherwise too little data will be collected. 

Also see https://data.syncthing.net/ for an example

Settings:
![image](https://user-images.githubusercontent.com/5652964/195068932-b99fca39-a620-47a5-b880-91973afdb919.png)

Preview:
![image](https://user-images.githubusercontent.com/5652964/195069103-cf989dc9-e2a1-4cf8-a02d-d4895ea7a058.png)

**Comment by ovizii** [comment]
> As long as PII is not being collected, this seems fine to me. I also don't see as opt-in being a viable option. If the goal of this is to gather data that guides Kopia development, it needs to be opt-out, because otherwise not much data will be gathered.

I totally agree.

**Comment by jkowalski** [comment]
Looks like syncthing has exactly the thing we would like to have here and roughly the same level of detail we want to collect (I don't care about geo location of clients). They have some great metrics I have not thought of, which would also be great to have here, in terms of hashing performance, upload speed, etc. whch optional features/policy settings are being used. I'm definitely going to include some those in final proposal and PR. I think there is potentially a ton of useful information to collect from feature development perspective, so we'll need to be selective to keep the ping size small (<1-2KB). Looks like they also send pings once every day. Great find @cristihcd!

@typpos Knowing this information will help decide things like:

- how much longer do we need to support v1 index format, which has been deprecated >1year ago. I know anecdotally some folks still force v1 format on repositories they create but I'd like to know how many that is
- how frequently des maintenance actually run and how long does it take for repository sizes? is there something to be done
- along the same lines, do we want to invest in live migration between formats?
- have we set the defaults for epoch manager correctly or can some parameters be optimized?
- are people using optimal hashing/compression/encryption algorithms for their platform? can we have better defaults or be smarter about this?
- how many ignore rules or compression exclusions do people have? can we help improve this experience?
- do folks regularly repositories with giant files (10-100s of GB) that likely does not scale well with existing default splitters?
- do we need to use a lot bigger blobs to reduce file count and save on maintenance time?

**Comment by jemail99** [comment]
Consent for telemetry must be clearly asked from the user at all stages eg. on first run, when ever the app is updated etc. Also with no default accept button already highlighted to push by accident, pushing enter will not advance the acceptance screen until a conscious choice has been made by the user. Until that conscious decision is made by the user telemetry is by default off.

You are producing a product that relies heavily on user trust. If you earn that trust your users will opt in.

**Comment by cristihcd** [comment]
just to add another datapoint, this is how timescaledb (database extension) presents telemetry:

```
WELCOME TO
 _____ _                               _     ____________
|_   _(_)                             | |    |  _  \ ___ \
  | |  _ _ __ ___   ___  ___  ___ __ _| | ___| | | | |_/ /
  | | | |  _ ` _ \ / _ \/ __|/ __/ _` | |/ _ \ | | | ___ \
  | | | | | | | | |  __/\__ \ (_| (_| | |  __/ |/ /| |_/ /
  |_| |_|_| |_| |_|\___||___/\___\__,_|_|\___|___/ \____/
               Running version 2.8.1
For more information on TimescaleDB, please visit the following links:

 1. Getting started: https://docs.timescale.com/timescaledb/latest/getting-started
 2. API reference documentation: https://docs.timescale.com/api/latest
 3. How TimescaleDB is designed: https://docs.timescale.com/timescaledb/latest/overview/core-concepts

Note: TimescaleDB collects anonymous reports to better understand and assist our users.
For more information and how to disable, please see our docs https://docs.timescale.com/timescaledb/latest/how-to-guides/configuration/telemetry.
```

**Comment by jemail99** [comment]
I don't think timescaledb is a good example here. Digging around docs on how to disable telemetry is deliberately making it harder for the user to make a choice. Effectively timescaledb are manufacturing consent for their telemetry. Any design choice which makes life harder for users will lose that users trust.

Issues of telemetry must be clear to the user without reference to anything beyond the screen they are reading. The user must be presented with information and a clear choice of opting in or opting out when telemetry is mentioned on screen. To do anything else is disingenuous and does not build trust.

**Comment by ntolia** [comment]
As with all things, there are tradeoffs. If we want meaningful data, I believe the answer has to be:
- Opt-in default
- Documentation pointer on what gets reported, how it is used, and how to turn telemetry off

Apart from the risk of a very leaky data funnel, putting prompts in is a poor UX and can break automation. 

Finally, I will say that based on my experience working in the backup industry, most products that are highly trusted by large and small customers do have some sort of telemetry and often very deep non-anonymous integration for commercial products. Folks that are really paranoid just run in "dark mode".

**Comment by WarmChocolateCake** [comment]
Just started looking at Kopia, so saw this as in interesting thread.

I use several products where Opt-In is the default. And I have Opted-In because:
1) I like the software
2) I want to continue liking the software :)

My largest observation is:

- Opt-In = High **Quality** of data (useful info for the developers)
- Opt-Out = High **Quantity** of data ("more" != "better")

In the future, if the devs want to stop, or change the data collected (Syncthing has changed their reports a few times), then the devs will need to deal with lots of data that still comes in, using the old format(s). _Generalisation_: lots of end-users won't change something that's "working" for years.

I also strongly agree with the other comments on knowing what data is being collected.

**Comment by jkowalski** [comment]
Additional thoughts I had:

We will have telemetry parameters defined as new type of policy with 3 fields:

* `consentLevel` - maximum version of telemetry collection to which the user has consented, initially will be `1`, in the future we may introduce 2, 3, ... which will require prompting the user again (0 == disabled, not set==ask the user)
* `endpoint` - address where telemetry is send, default `telemetry.kopia.io:443` but can be overridden to allow internal collection of telemetry (we'll publish the source code of this along with proxy mode which lets companies capture their own users' telemetry but still share it with Kopia upstream)
* `telemetryID` - random identifier of the repository


For new repositories we will set default `global` policy to either:

```
{
  "telemetry": {
    "consentLevel": 1,
    "endpoint": "telmetry.kopia.io:443",
    "telemetryID": <random>,
  }
}
```

or

```
{
  "telemetry": {
    "consentLevel": 0
  }
}
```

Now because those settings are policy-based they will be merged as usual with inheritance, it's possible to override a single user or host, for example to exclude them from telemetry or their data to another endpoint:

```
$ kopia policy set --global --telemetry-consent 1

# disable consent for one user
$ kopia policy set user@host --telemetry-consent 0

# all users on a given host will send their telemetry somewhere else
$ kopia policy set @host --telemetry-endpoint internal-telemetry:1234
```

**Comment by WarmChocolateCake** [comment]
> * `endpoint` - address where telemetry is send, default `telemetry.kopia.io:443` but can be overridden to allow internal collection of telemetry [...]

This is an excellent idea. Several systems I work on are totally air-gapped, so having our own telemetry would be great. In my particular case, that would mean having to store & transfer the telemetry to kopia.io rather than proxying it, but I guess that's a real edge case...

[...]
> For new repositories we will set default global policy to [...]

[...]

> or
> 
> ```
> {
>   "telemetry": {
>     "consentLevel": 0
>   }
> }
> ```

This one 👆  🙂

**Comment by jkowalski** [comment]
Yeah, @WarmChocolateCake this will be based on user explicit decision, no default here.

**Comment by jkowalski** [comment]
I started implementing the consent-gathering UX, here's the text I'm planning to use. Comments welcome.

The following dialog will be used in:

* CLI when running `kopia repository create` after password prompt
* CLI when running as interactive user when the consent has not been decided yet
* UI when opening the window for the repository when the consent has not been decided yet

```
Please read carefully:

Kopia would like to request your permission to periodically share anonymous
statistics about this repository and crash reports with project developers.

This information includes:

- version of Kopia, hardware architecture and operating system information
- repository format (encryption, compression, hash algorithms, etc.)
- basic statistics about repository (such as # files, average sizes, etc.)
- performance metrics (such as upload, compression speed, memory usage, etc.)
- information about which features of Kopia are most frequently used

We DO NOT collect any information that could be used to identify users,
such as usernames, file names, or hostnames.

See https://kopia.io/docs/advanced/telemetry/ for more information.

[1] Yes, I'm ok to share anonymous data about this repository.
[2] No, I prefer not to share. Don't ask me again for this repository.
[3] Maybe later, I'm not ready to decide yet, ask me again in a week.
[4] I'd like to see the data that would be shared.

Enter 1-4: 
```

**Comment by basldfalksjdf** [comment]
Perhaps remove "please read carefully" at the beginning. It makes the message sound very ominous. They have to read and respond anyway, I don't think the "please read carefully" is necessary.

Also, this `- repository format (encryption, compression, hash algorithms, etc.)` may make users think that encryption password is being transmitted. So maybe say `- repository format (type of encryption, compression, hash algorithms, etc.)` and then say `We do NOT collect encryption passwords, and we do NOT collect any information that could be used to identify users or files, such as IP addresses, usernames, file names, or hostnames.`

**Comment by jkowalski** [comment]
@basldfalksjdf updated text above.

**Comment by jkowalski** [comment]
I've been thinking about how and where to actually collect the statistics and it's less than obvious, given that we want to send them every 24 hours. 

I have an idea that would actually be a cool feature in itself regardless of telemetry, namely ability to store basic monitoring data for kopia clients over time in the repository itself. The data collected this way will be shared through telemetry endpoint - so killing two birds with one stone.

1. Each kopia client will collect runtime performance stats (counters, gauges and histograms) from each run in local files `<some-dir>/yyyy-MM-dd/<uuid>` that will be periodically be saved to disk (on every run and every hour for long-running processes)

2. Periodically (every 24 hours) all local files will be aggregated into one per-day set of statistics and saved in a repository as a manifest of some new type (`client-stats`). All local files from previous days will be deleted.

3. **We will send the exact same data to the telemetry endpoint.** so users will have full visibility into what's in those telemetry reports.

4. Stats in the repository for each client will be kept for a configurable amount number of days, initiallly 30 and pruned during maintenance.

5. Later we can develop some CLI commands and perhaps even plot some nice graphs in the UI to show some key statistics and maybe export data for analysis.

**Comment by basldfalksjdf** [comment]
> @basldfalksjdf updated text above.

Looks good. The only one thing I would change further is remove the comma here: `We do NOT collect encryption password, or any information...`

**Comment by basldfalksjdf** [comment]
@jkowalski: That does sound cool, but my only one question is -- would this be adding a significant amount of unnecessary complexity to Kopia and repositories? Could it add another potential point of failure that may corrupt backups?

**Comment by jkowalski** [comment]
Yeah, complexity of this is to be seen, hopefully simple refactoring to get to this point. That's why i'm moving very slowly on this whole idea, trying to carefully weight the complexity vs benefits.

**Comment by stpr-dev** [comment]
> I've been thinking about how and where to actually collect the statistics and it's less than obvious, given that we want to send them every 24 hours.

Is there a necessity to send it that frequently? As in, do you need that level of resolution? 24H seems too frequent, it might be better to be a bit more sporadic (I was thinking weekly or something along those lines). If you are planning to aggregate everything anyway, would it not be better to aggregate data and send it more sporadically?

**Comment by stpr-dev** [comment]
Also, @jkowalski, it might be a good idea to pin this discussion given the amount of engagement going on. I would also suggest perhaps enabling the Discussions feature of GitHub and move it there. There are a lot of questions in the issues tab that could be part of the discussions page, just to keep the issues tab focused on technical issues.

**Comment by WarmChocolateCake** [comment]
Agree that this should be a "Discussion"...

Back to the point on storing the telemetry data ...

I like that idea because (if I've understood correctly) it gives me the opportunity to look back at "last night's" statistics - if we always transfer a similar quantity of data and something changes, then I can investigate earlier, rather than automated backups running for weeks and not doing what I think they're doing.

(I'm new to Kopia, so if that data is already stored elsewhere, I'm just not aware yet...)

**Comment by Shrekster** [comment]
What is th data expiration scheme of the local aggregator. I read this:

> ... individual ping records will be discarded after no more than 90 days with only aggregated counters stored beyond that.

I suppose the idea is that the data expiration scheme would also be part of the telemetry aggregator+proxy code that will either be hosted on-prem or the default `telemetry.kopia.io:443`. Or is the idea that this would be supported on the default telemetry backend only ?

**Comment by pauby** [comment]
For me, the answer to this is simple. What is your priority? 

If your priority is to gather telemetry data, then you need to make this an opt-out default. That is not consent.

If your priority is the privacy of your users, then you need to make it an opt-in default. That is consent.

You cannot have both telemetry and privacy as a priority because they conflict.

There was [mention here that an opt-in wouldn't provide as much telemetry](https://github.com/kopia/kopia/issues/2486#issuecomment-1267998925). _Some_ telemetry data is better than what you have now, and be good enough to make product decisions on? Would it not be wise to try that and then revisit this later? Doing that allows people to give you their trust rather than you taking it.

**Comment by tiagocpontesp** [comment]
why not embed a poll in the documentation?
could further differentiate intended uses from actual ones.

**Comment by kronenpj** [comment]
May I also add a suggestion to ask the user for consent whenever the category of data collected changes. E.g. if information moves from specific values to ranges or vice-versa. Collecting a new type data like compression ratio that wasn't collected before. 

I've heard of companies that have encountered a problem in the past where telemetry collection was opt-out and initially benign. A software update started sending sensitive information without indicating a change had been made to what it collects to give the opportunity to understand the new collection and opt-out.

Transparency will be key in getting users to opt-in or not opt-out. Yes, this is an open-source project, but not everyone will want or be able to dig through the code to find what might have changed.
